src/main/scala/com/daml/projection/javadsl/BatchSource.scala

```diff
@@ -112,4 +134,57 @@ object BatchSource {
           J.TreeEvent.fromProtoTreeEvent(ST.TreeEvent.toJavaProto(event))))
     }
   }.toJava
+
+  @FunctionalInterface
+  trait GetContractTypeId[E] {
+    def fromEvent(event: E): ju.Optional[J.Identifier]
+    def toScala: SGetContractTypeId[E] = (event: E) =>
+      fromEvent(event).toScala.map(i => Identifier.fromJavaProto(i.toProto))
+  }
+
+  object GetContractTypeId {
+    implicit val `from event`: GetContractTypeId[J.Event] = fromEvent
+    def fromEvent: GetContractTypeId[J.Event] = (event: J.Event) => {
```

Comment:
```suggestion
    def fromEvent(): GetContractTypeId[J.Event] = (event: J.Event) => {
```
For Java usage so non-nullary please for all of these.